### PR TITLE
Extract FPD Merge To Separate Package

### PR DESCRIPTION
**Pull request by SyntaxNode**
Pulling out this code for potential reuse and to isolate this behavior from FPD logic. Changed the function signature to be a pure method, but is functionality identical.

Realized some of the clone methods are expensive and most are likely not necessary for the merge. This will get worse as I look to introduce an imp merge function as the imp structure is much deeper.

I'm going to investigate a new merge approach which will only clone data present in the json override. Need to benchmark if the lazy cloner would be overall more efficient. Certainly would produce a lot less allocations, just need to verify the CPU cost isn't much higher.

### CODE REVIEW

File: ortb/merge/app_test.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			expectError: true,
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			originalApp := ortb.CloneApp(&test.givenApp)
+			merged, err := App(&test.givenApp, test.givenJson)
+
+			assert.Equal(t, &test.givenApp, originalApp)
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply ortb/merge/app_test.go:178]: Checked, app site and user look right!

**Review by VeronikaSolovei9 — APPROVED** [review]

File: firstpartydata/first_party_data.go [hunk 2]
```diff
@@ -195,42 +199,19 @@ func resolveUser(fpdConfig *openrtb_ext.ORTB2, bidRequestUser *openrtb2.User, gl
...
 	}
 	if fpdConfigUser != nil {
-		if err := mergeUser(newUser, fpdConfigUser); err != nil {
+		if mergedUser, err := merge.User(newUser, fpdConfigUser); err != nil {
+			if err == merge.ErrBadOverride {
+				return nil, ErrBadFPD
+			}
 			return nil, err
+		} else {
+			newUser = mergedUser
```
Comment (guscarreon) [inline firstpartydata/first_party_data.go:208]: When `err != nil` evaluates to `true` we end up returning `nil` in both branches, do we really need `mergedUser` as a separate variable? We could even spare the `else` clause:
```
201     if fpdConfigUser != nil {
202 -       if mergedUser, err := merge.User(newUser, fpdConfigUser); err != nil {
    +       if newUser, err := merge.User(newUser, fpdConfigUser); err != nil {
203             if err == merge.ErrBadOverride {
204                 return nil, ErrBadFPD
205             }
206             return nil, err
207 -       } else {
208 -           newUser = mergedUser
209         }
210     }
firstpartydata/first_party_data.go
```
But I see this pattern repeated in other places below (`mergedSite` and `mergedApp`) should we modify to make it leaner? Also, why do `merge.User`, `merge.Site` and `merge.App` are now returning pointers? Is it to check for shared memory issues?

Comment (SyntaxNode) [reply firstpartydata/first_party_data.go:208]: > But I see this pattern repeated in other places below (mergedSite and mergedApp) should we modify to make it leaner? 

Yes. My original intent was to not assign mergedUser if there is an error, but if there is an error there is no return path for newUser. Updated.

> Also, why do merge.User, merge.Site and merge.App are now returning pointers?

They always did this. Before, they changed the memory address of the input pointer. Now it returns a separate pointer and leaves the input alone.

**Review by guscarreon — APPROVED** [review]
LGTM

**Pull request merged**